include/mfmg/dealii/lanczos.templates.hpp

```diff
@@ -0,0 +1,494 @@
+      }
+    }
+  }
+
+#if 0
+
+void LAPACK_dgeev( char* jobvl, char* jobvr, lapack_int* n, double* a,
+                   lapack_int* lda, double* wr, double* wi, double* vl,
+                   lapack_int* ldvl, double* vr, lapack_int* ldvr, double* work,
+                   lapack_int* lwork, lapack_int *info );
+
+lapack_int LAPACKE_dgeev( int matrix_order, char jobvl, char jobvr,
+                          lapack_int n, double* a, lapack_int lda, double* wr,
+                          double* wi, double* vl, lapack_int ldvl, double* vr,
+                          lapack_int ldvr );
+
+lapack_int LAPACKE_dgeev_work( int matrix_order, char jobvl, char jobvr,
+                               lapack_int n, double* a, lapack_int lda,
+                               double* wr, double* wi, double* vl,
+                               lapack_int ldvl, double* vr, lapack_int ldvr,
+                               double* work, lapack_int lwork );
+
+
+http://www.netlib.org/lapack/explore-html/d9/d8e/group__double_g_eeigen_ga66e19253344358f5dee1e60502b9e96f.html#ga66e19253344358f5dee1e60502b9e96f
+#endif
+}
+
+//-----------------------------------------------------------------------------
+/// \brief Lanczos solver: perform convergence check
+
+template <typename OperatorType, typename VectorType>
+bool Lanczos<OperatorType, VectorType>::details_check_convergence(
+    ScalarType beta, const int num_requested, double tol,
+    std::vector<ScalarType> const &evecs)
+{
+  const int n = evecs.size();
+
+  // Must iterate at least until we have num_requested eigenpairs.
+  if (n < num_requested)
+  {
+    return false;
+  }
+
+  bool is_converged = true;
+
+  if (_verbosity > 0)
+  {
+    std::cout.precision(4);
+    std::cout << std::fixed;
+    // std::cout << "   " << beta;
+    std::cout << "   bounds ";
+  }
+
+  // Terminate if every approx eval has converged to tolerance
+  // ISSUE: here ignoring possible nuances regarding the correctness
+  // of this check.
+  // NOTE: k may be desirable to "scale" the stopping criterion
+  // based on (estimate of) matrix norm or similar.
+  for (int i = 0; i < num_requested; ++i)
+  {
+    const double bound = (double)beta * abs((double)evecs[n - 1 + n * i]);
+    is_converged = is_converged && bound <= tol;
+    if (_verbosity > 0)
+    {
+      std::cout << " " << bound;
+    }
+  }
+
+  if (_verbosity > 0)
+  {
+    std::cout << "\n";
+  }
+
+  return is_converged;
+}
+
+//-----------------------------------------------------------------------------
+/// \brief Lanczos solver: calculate full (approx) evecs from tridiag evecs
+template <typename OperatorType, typename VectorType>
+void Lanczos<OperatorType, VectorType>::details_calc_evecs(
+    const int num_requested, const int n,
+    std::vector<VectorType> const &lanc_vectors,
+    std::vector<ScalarType> const &evecs_tridiag,
+    std::vector<VectorType> &evecs)
+{
+  assert(evecs.size() == 0);
```

Comment:
```suggestion
  assert(evecs.empty());
```
if you really want the assertion.